apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1520,78 +1568,70 @@ void RetryAcksAndExtends(bool? ack, AckError ackError, IEnumerable<TimedId> exte
-                }
-
-                // Updates the backoff of the retryableErrorIds in _retryableIds dictionary.
-                void UpdateAllRetryableErrorBackoff(IEnumerable<string> ids, TimeSpan? backoff)
-                {
-                    foreach (var id in ids)
+                    // The ids that exist in both retryableErrorIds and _retryableIds now are the ones that need to be retried.
+                    // We can have ids with different back-off.
+                    // Group these ids on back-off and add them to the corresponding queue in the increasing order of backoff
+                    foreach (var retryGroup in _retryableIds.Where(pair => retryableErrorIds.Contains(pair.Key)).GroupBy(pair => pair.Value.Backoff).OrderBy(pair => pair.Key))
                     {
-                        if (_retryableIds.TryGetValue(id, out var info))
-                        {
-                            _retryableIds[id] = info.WithBackoff(backoff);
-                        }
+                        var backoff = retryGroup.Key ?? TimeSpan.Zero;
+                        var retryIds = retryGroup.Select(j => j.Key);
+                        Task delayTask = _scheduler.Delay(backoff, _softStopCts.Token);
+                        Add(delayTask, new NextAction(false, hasAcksOrNacks
+                        ? () => { ackActionToRetry(retryIds); StartPush(); }
+                        : () => { extendActionToRetry(extendIds.Where(j => retryIds.Contains(j.Id))); StartPush(); }));
                     }
                 }
 
-                // Updates the _retryableIds dictionary. Ensures that the ids that should be retried exist in the dictionary.
-                // By design, only ids that are in _retryableIds dictionary will be retried with exponential backoff.
+                // Updates the _retryableIds dictionary. Ensures that the ids that should be re-tried exist in the dictionary.
+                // By design, only ids that are in _retryableIds dictionary will be retried with exponential back-off.
                 // Removes the ids that are either successful or become permanent failures as they should not be retried.
+                // allIds - All the ids for which Acknowledge/ModifyAckDeadline request was sent.
+                // ackError - The AckError that occurred in the response.
                 void UpdateRetryableIds(IEnumerable<string> allIds, AckError ackError)
                 {
                     var temporaryFailureIds = ackError?.TemporaryFailureIds ?? Enumerable.Empty<string>();
                     var permanentFailureIds = ackError?.PermanentFailureIds ?? Enumerable.Empty<string>();
                     // Successful Ids = AllIds - (TemporaryFailures + Permanent failures)
-                    var successfulIds = allIds.Locked(() => allIds.Except(temporaryFailureIds.Union(permanentFailureIds)));
+                    var successfulIds = allIds.Except(temporaryFailureIds).Except(permanentFailureIds);
```

Comment:
We should consider the impact of lazy evaluation at some point - I suspect we're fine, but it might be worth adding a TODO.

Reply:
Done.